src/pages/blog/2024-06-28-why-i-like-graphql.mdx

```diff
@@ -0,0 +1,92 @@
+
+For a public API, a coarser-grained, resource-based API works great, and can be described through OpenAPI. SDKs can be generated through amazing tools like [Kiota](https://learn.microsoft.com/en-us/openapi/kiota/overview). It's hard to beat a well-made SDK for a public API, and in my experience, that's actually what customers expect and want to use.
+
+## Authorization
+
+Authorization is a challenge with GraphQL. The thing is it's almost always challenging no matter what API style you use. I'd go as far as saying it is a challenge with designing software in general. Here's an example from a recent post that highlights this very well:
+
+```graphql
+query {
+  user(id: 321) {
+    handle # ✅ I am allowed to view Users public info
+    email # 🛑 I shouldn't be able to see their PII just because I can view the User
+  }
+  user(id: 123) {
+    blockedUsers {
+      # 🛑 And sometimes I shouldn't even be able to see their public info,
+      # because context matters!
+      handle
+    }
+  }
+}
+```
+
+`handle` and `email` both have different authorization rules. This is actually quite tricky to handle with a `GET /user/:id` endpoint as well. There's really nothing that makes GraphQL harder here. Yes when fine-grained authorization is needed, you'll need fine-grained authorization checks at that level.
+
+```graphql
+user(id: 123) {
+  blockedUsers {
+    # 🛑 And sometimes I shouldn't even be able to see their public info,
+    # because context matters!
+    handle
+  }
+}
+```
+
+This part is interesting as well and another challenge of authorizing code and models in general. The same "object" accessed through different contexts can actually have different authorization rules. Again, this is common in all API styles as well. That's why I actually usually advise designing this through a different model entirely, since it's likely they'll evolve differently. Here this is even possibly an API design mistake instead. If `handle` should never be seen for `blockedUsers`, then it shouldn't even be part of the schema here. This is a super common mistake where folks try to reuse common models/types instead of being specific.
+
+After 8 years of GraphQL, I realize more and more that authorization is much more than a GraphQL problem. The API layer part is easy, but most code-bases are much more complex and must guard against unauthorized access in much better ways. Companies like [oso](https://www.osohq.com/) and [authzed](https://authzed.com/) and two great examples of how to do this well but also how complex this thing can be in general.
```

Comment:
I would add here that this is exactly why graphql.org best practices guide has **always** suggested not to put Authorization as part of your GraphQL layer:
https://graphql.org/learn/authorization/